### PR TITLE
feat: monitoring for completeness of state dumps

### PR DESCRIPTION
**Pull request by ppca**
This tool enables the following:
1. you can do a single check of the completeness(all parts dumped) and validity (all parts valid) of an epoch's state parts by `./target/debug/neard state-parts-dump-check --chain-id mainnet --gcs-bucket state-parts single-check --shard-id {shard_id} --epoch-id {epoch_id} --epoch-height {epoch_height} --state-root {state_root}`
2. you can run the command to check all 4 shards' dumps endlessly by `./target/debug/neard state-parts-dump-check --chain-id mainnet --gcs-bucket state-parts loop-check`: This will poll for latest epoch -1, and once it updates, the tool will check all 4 shards of that epoch's dump. Note you can also use this to check your betanet or mocknet dumps, the only thing different is you need to supply the address (http://{ip_addr}:{port}) of your rpc node in that network: `./target/debug/neard state-parts-dump-check --chain-id mocknet --gcs-bucket state-parts loop-check --rpc-server-addr {rpc_server_addr}`

How does the tool work?
If you run the single check, it would use the epoch_id, epoch_height, network, and storage location to download the state parts, and use the state root you supplied to validate them.
If you run the loop check, the logic is an infinite loop that polls for latest epoch, once there's a new epoch, it will start downloading and validating for that epoch. The 4 shards run one after another. 

At each new epoch check, the tool will emit metrics for number of total required parts, number of dumped parts, number of valid/invalid parts, epoch_height. If the new check was more than 1 height higher than last checked epoch, we emit 1 for skipped epoch metric. If the app is running, we will emit 1 for near_state_sync_dump_check_process_is_up.

### CODE REVIEW

File: tools/state-parts-dump/src/rpc_requests.rs [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+}
+
+pub struct RpcClient {
+    client: reqwest::blocking::Client,
+    rpc_endpoint: String,
+}
+
+impl RpcClient {
+    // network can only be "mainnet" or "testnet"
+    pub fn new(network: &str) -> Self {
```
Comment (nikurt) [inline tools/state-parts-dump/src/rpc_requests.rs:61]:
```suggestion
    pub fn new(chain_id: &str) -> Self {
```

File: tools/state-parts-dump/src/rpc_requests.rs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+                "jsonrpc": "2.0",
+                "id": "dontcare",
+                "method": "validators",
+                "params": {
+                    "epoch_id": epoch_id
+                }
+            }))
+            .send()?
+            .json()?;
+        let epoch_start_height = epoch_response.result.epoch_start_height;
```
Comment (nikurt) [inline tools/state-parts-dump/src/rpc_requests.rs:99]: nice, I always forget about this RPC as its epoch_id parameter is not documented https://docs.near.org/api/rpc/network#validation-status

File: tools/state-parts-dump/src/rpc_requests.rs [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+                }
+            }))
+            .send()?
+            .json()?;
+        let chunks = prev_epoch_last_block_response.result.chunks;
+        let state_root = &chunks[chunks.len() -1].prev_state_root;
+        Ok(state_root.to_string())
+    }
+
+    pub fn get_latest_block_hash(&self) -> Result<String, Box<dyn Error>> {
```
Comment (nikurt) [inline tools/state-parts-dump/src/rpc_requests.rs:162]: As it's querying the final block and not the latest block, should this be named `get_final_block_hash()`?

```suggestion
    pub fn get_final_block_hash(&self) -> Result<String, Box<dyn Error>> { 
```

File: tools/state-parts-dump/src/cli.rs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+                crate::metrics::STATE_SYNC_DUMP_CHECK_NUM_PARTS_INVALID.with_label_values(&[&self.shard_id.to_string()]).inc();
+                println!("part {} is invalid", part_id);
+            }
+        };
+        Ok(StatePartsDumpCheckStatus::Done { epoch_height: self.epoch_height })
+    }
+}
+
+
+pub fn create_external_connection(
```
Comment (nikurt) [inline tools/state-parts-dump/src/cli.rs:124]: I've seen this function before somewhere :eyes:

File: tools/state-parts-dump/src/cli.rs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+            self.s3_bucket.clone(),
+            self.s3_region.clone(),
+            self.gcs_bucket.clone(),
+        );
+
+        let directory_path = external_storage_location_directory(&self.chain_id, &self.epoch_id, self.epoch_height, self.shard_id);
+        let part_file_names = external.list_state_parts(self.shard_id, &directory_path).await?;
+        // TODO: return error instead of assert!()
+        assert!(!part_file_names.is_empty(), "Number of state parts is 0");
+        let num_parts = part_file_names.len() as u64;
```
Comment (nikurt) [inline tools/state-parts-dump/src/cli.rs:77]: Imagine a directory contains only the following files:
* `part-000000-of-000002`
* `part-000000-of-000003`

The code will see `num_parts = 2` and `total_required_parts = 2` and be happy about the outcome.

File: tools/state-parts-dump/src/cli.rs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+            return Ok(StatePartsDumpCheckStatus::WaitingForParts { epoch_height: self.epoch_height })
+        } else if num_parts > total_required_parts{
+            println!("total state parts required: {} > number of parts already dumped: {}, there are more dumped parts than total required, something is seriously wrong", total_required_parts, num_parts);
+            return Ok(StatePartsDumpCheckStatus::Done { epoch_height: self.epoch_height })
+        }
+
+        // TODO: run in parallel with rayon
+        for part_id in 0..total_required_parts {
+            if part_id == 0 {
+                crate::metrics::STATE_SYNC_DUMP_CHECK_NUM_PARTS_VALID.with_label_values(&[&self.shard_id.to_string()]).set(0);
```
Comment (nikurt) [inline tools/state-parts-dump/src/cli.rs:94]: This should be done by the code that updates `self.shard_id`.

File: tools/state-parts-dump/src/cli.rs [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+                println!("part {} is invalid", part_id);
+            }
+        };
+        Ok(StatePartsDumpCheckStatus::Done { epoch_height: self.epoch_height })
+    }
+}
+
+
+pub fn create_external_connection(
+    root_dir: Option<PathBuf>,
```
Comment (nikurt) [inline tools/state-parts-dump/src/cli.rs:125]: Why do you support `root_dir`?

File: tools/state-parts-dump/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+    let rpc_client = crate::rpc_requests::RpcClient::new("mainnet");
+    
+    let mut last_check_status = Ok(StatePartsDumpCheckStatus::WaitingForParts{epoch_height: 0});
+
+    let mut last_check_iter_info;
+    while true {
+        let dump_check_iter_info_res = get_processing_epoch_information(&rpc_client);
+        if let Err(_) = dump_check_iter_info_res {
+            break;
```
Comment (nikurt) [inline tools/state-parts-dump/src/main.rs:24]: Expect RPCs to fail regularly and retry the errors (maybe with increasing intervals between retries).

File: tools/state-parts-dump/src/rpc_requests.rs [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+    prev_state_root: String,
+}
+
+pub struct RpcClient {
+    client: reqwest::blocking::Client,
+    rpc_endpoint: String,
+}
+
+impl RpcClient {
+    // network can only be "mainnet" or "testnet"
```
Comment (nikurt) [inline tools/state-parts-dump/src/rpc_requests.rs:60]: Should be fine to accept the rpc_endpoint as an CLI parameter.
This doesn't seem bad to me:
`state-parts-dump --rpc-endpoint https://rpc.mainnet.near.org`

File: Cargo.toml [hunk 10]
```diff
@@ -77,6 +77,7 @@ members = [
     "tools/rpctypegen/macro",
     "tools/speedy_sync",
     "tools/state-parts",
+    "tools/state-parts-dump",
```
Comment (nikurt) [inline Cargo.toml:80]: How about `tools/state-dump-monitoring`?

File: tools/state-parts-dump/src/main.rs [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+        }
+        let dump_check_iter_info = dump_check_iter_info_res.unwrap();
+        
+        if let Ok(StatePartsDumpCheckStatus::Done{epoch_height: epoch_height}) = last_check_status && epoch_height >= dump_check_iter_info.prev_epoch_height {
+            // TODO: sleep 5 mins
```
Comment (nikurt) [inline tools/state-parts-dump/src/main.rs:29]: IIUC, the code checks the dumps of one epoch and if that epoch is done, it `break`s and terminates.
Shouldn't the code check state dumps for the next epochs?

Comment (nikurt) [inline tools/state-parts-dump/src/main.rs:29]: TODO: Check that no epoch_heights were skipped.

File: tools/state-parts-dump/src/helpers.rs [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+// TODO: run all 4 shards in parallel
+pub(crate) async fn run_dump_check(
+    rpc_client: &crate::rpc_requests::RpcClient,
+    dump_check_iter_info: &DumpCheckIterInfo,
+) -> anyhow::Result<crate::cli::StatePartsDumpCheckStatus> {
+    println!( "start");
+    let state_root_str = rpc_client.get_prev_epoch_state_root(&dump_check_iter_info.prev_epoch_id).or_else(anyhow!("get_prev_epoch_state_root failed"))?;
+
+    let state_root: StateRoot = CryptoHash::from_str(&state_root_str)?;
```
Comment (nikurt) [inline tools/state-parts-dump/src/helpers.rs:39]: Consider changing return types of your helper functions to be `CryptoHash` and `EpochId`.

File: tools/state-parts-dump/src/rpc_requests.rs [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+                    "block_id": prev_epoch_last_block_hash
+                }
+            }))
+            .send()?
+            .json()?;
+        Ok(prev_epoch_last_block_response.result.header.epoch_id)
+    }
+
+    // get the state root of previous epoch. This should be the state root of the last block of last epoch.
+    pub fn get_prev_epoch_state_root(&self, block_hash: &String) -> Result<String, Box<dyn Error>> {
```
Comment (nikurt) [inline tools/state-parts-dump/src/rpc_requests.rs:142]: Oh, you get state root from the block RPC method and not from `StateRequestHeader`.
And you don't need the peer-to-peer connections in this tool.

Nice!
The less coupling the better.

File: tools/state-parts-dump/src/rpc_requests.rs [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+            .send()?
+            .json()?;
+        let block_hash = block_response.result.header.hash;
+
+        Ok(block_hash)
+    }
+
+    // epoch_id here is the epoch_id you want state part for
+    pub fn get_current_epoch_first_block_hash(&self, epoch_id: &String) -> Result<String, Box<dyn Error>> {
+        let epoch_response: ValidatorsResponse = self
```
Comment (nikurt) [inline tools/state-parts-dump/src/rpc_requests.rs:86]: Use this instead: https://github.com/near/nearcore/blob/1e41344e37f10b83460fbaf1cd0ffa2c79e8fb5d/chain/jsonrpc/client/src/lib.rs#L193

And please extend that method to also accept `EpochId` and not only `MaybeBlockId` :) The community will love that.

File: tools/state-parts-dump/src/cli.rs (hunk 4 above)
Comment (ppca) [reply tools/state-parts-dump/src/cli.rs:124]: This was in state sync code but I removed some unnecessary arguments thus a new one here

File: tools/state-parts-dump/src/cli.rs (hunk 5 above)
Comment (ppca) [reply tools/state-parts-dump/src/cli.rs:77]: Ah nice catch! I'll modify

File: tools/state-parts-dump/src/main.rs (hunk 11 above)
Comment (ppca) [reply tools/state-parts-dump/src/main.rs:29]: Ah I meant for it to sleep and continue. This break is merely for me to see if it succeeds just for one run.

File: tools/state-parts-dump/src/helpers.rs (hunk 12 above)
Comment (ppca) [reply tools/state-parts-dump/src/helpers.rs:39]: hmmm keeping them as strings make it easier to use as input to other rpc request functions. But yeah if we want to directly use the result better to convert them. hmm

**Comment by ppca** [comment]
modified per our last discussion @nikurt 
Still need to add logic to process 4 shards in parallel and test on a node

File: nearcore/src/state_sync.rs [hunk 15]
```diff
@@ -138,7 +138,7 @@ impl StateSyncDumpHandle {
     }
 }
 
-fn extract_part_id_from_part_file_name(file_name: &String) -> u64 {
+pub fn extract_part_id_from_part_file_name(file_name: &String) -> u64 {
```
Comment (nikurt) [inline nearcore/src/state_sync.rs:141]:
```suggestion
pub fn extract_part_id_from_part_file_name(file_name: &str) -> u64 {
```

File: tools/state-parts-dump-check/Cargo.toml [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+repository.workspace = true
+license.workspace = true
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+actix.workspace = true
+borsh.workspace = true
+clap.workspace = true
+reqwest = { version = "0.11.6", features = ["blocking", "json"]  }
```
Comment (nikurt) [inline tools/state-parts-dump-check/Cargo.toml:16]: Please keep definitions of all dependencies in the root `Cargo.toml`.

File: tools/state-parts-dump-check/src/rpc_requests.rs [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+struct ChunkInfo {
+    prev_state_root: String,
+}
+
+pub struct RpcClient {
+    client: reqwest::blocking::Client,
+    rpc_endpoint: String,
+}
+
+impl RpcClient {
```
Comment (nikurt) [inline tools/state-parts-dump-check/src/rpc_requests.rs:43]: Please try using `chain/jsonrpc/client/src/lib.rs` before writing your own client.

File: tools/state-parts-dump-check/src/cli.rs [hunk 18]
```diff
@@ -0,0 +1,471 @@
...
+    gcs_bucket: Option<String>,
+    shard_id: ShardId,
+    rpc_client: crate::rpc_requests::RpcClient,
+) -> anyhow::Result<()> {
+    let mut last_check_status =
+        Ok(StatePartsDumpCheckStatus::WaitingForParts { epoch_height: 0 });
+
+    let mut last_check_iter_info;
+
+    loop {
```
Comment (nikurt) [inline tools/state-parts-dump-check/src/cli.rs:237]: Stopping and starting actix repeatedly may be problematic, but I don't know why.
I suggest to start and stop actix only once, here, and not on line 282.

```suggestion
 sys.block_on(async move {
    loop {
```

**Comment by ppca** [comment]
For each epoch, 4 shards check in sequence. Within each shard, parts are downloaded and validated in parallel thru tokio::spawn(), which basically create tasks to put into a thread pool with threads = #cores.
Also exposed prometheus at 9000 port /metrics endpoint.
Next up:
1) make grafana dashboard
2) switch out my own rpc client with existing `chain/jsonrpc/client/src/lib.rs`

**Review by nikurt — COMMENTED** [review]
Please add comments to all new public structs and functions.

**Pull request merged**